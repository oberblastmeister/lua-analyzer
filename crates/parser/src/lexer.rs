use std::str::Chars;

use rowan::TextSize;

use crate::{SyntaxKind, T};

pub(crate) const EOF_CHAR: char = '\0';

/// A lua token
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Token {
    /// The kind of token.
    pub kind: SyntaxKind,
    /// The length of the token.
    pub len: TextSize,
}

impl Token {
    fn new(kind: SyntaxKind, len: u32) -> Token {
        Token {
            kind,
            len: TextSize::from(len),
        }
    }
}

pub struct Lexer<'a> {
    input_len: u32,
    chars: Chars<'a>,
}

impl<'a> Lexer<'a> {
    fn new(input: &'a str) -> Lexer<'a> {
        Lexer {
            input_len: input.len() as u32,
            chars: input.chars(),
        }
    }

    fn chars(&self) -> Chars<'a> {
        self.chars.clone()
    }

    fn nth(&self, n: usize) -> char {
        self.chars().nth(n).unwrap_or(EOF_CHAR)
    }

    /// Peeks next char from stream without consuming it
    fn peek(&self) -> char {
        self.nth(0)
    }

    /// Checks if there is nothing more to consume.
    fn is_eof(&self) -> bool {
        self.chars.as_str().is_empty()
    }

    fn chars_len(&self) -> u32 {
        self.chars.as_str().len() as u32
    }

    /// Returns amount of already consumed symbols.
    fn pos(&self) -> u32 {
        self.input_len - self.chars_len()
    }

    fn bump(&mut self) -> Option<char> {
        self.chars.next()
    }

    fn bump_peek(&mut self) -> char {
        self.bump();
        self.peek()
    }

    fn lex_main(&mut self) -> SyntaxKind {
        let c = self.peek();

        let kind = match c {
            '(' => T!['('],
            ')' => T![')'],
            '{' => T!['{'],
            '}' => T!['}'],
            '[' => T!['['],
            ']' => T![']'],

            ',' => T![,],
            '.' => T![.],
            ';' => T![;],
            ':' => T![:],

            '+' => T![+],
            '*' => T![*],
            '/' => T![/],

            '!' => T![!],

            '-' => match self.bump_peek() {
                '-' => self.comment(),
                _ => T![-],
            },

            '\0' => T![eof],

            _ => T![unknown],
        };

        kind
    }

    fn comment(&mut self) -> SyntaxKind {
        assert_eq!(self.peek(), '-');
        self.chars.find(|c| *c == '\n');
        T![comment]
    }
}
