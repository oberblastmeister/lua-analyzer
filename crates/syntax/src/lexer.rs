use std::str::Chars;
use std::{iter, ops::Range};

use itertools::{Either, Itertools};
use rowan::{TextRange, TextSize};
use thiserror::Error;

use crate::{SyntaxError, SyntaxKind, T};
use parser::Token;

pub(crate) const EOF_CHAR: char = '\0';

macro_rules! done {
    ($expr:expr) => {
        return Ok($expr);
    };
}

macro_rules! assert_matches {
    ($expr:expr, $($stuff:tt)+) => {
        assert!(matches!($expr, $($stuff)+))
    };
}

struct WithLen<T> {
    inner: T,
    len: TextSize,
}

impl<T> WithLen<T> {
    fn new(t: T, len: TextSize) -> Self {
        Self { inner: t, len }
    }
}

pub fn tokenize(text: &str) -> (Vec<Token>, Vec<SyntaxError>) {
    let mut tokens = vec![];
    let mut errors = vec![];
    for res in tokenize_iter(text) {
        match res {
            Ok(tok) => tokens.push(tok),
            Err(e) => {
                tokens.push(e.to_unknown_token());
                errors.push(e.into());
            }
        }
    }
    (tokens, errors)
}

pub fn tokenize_iter(mut input: &str) -> impl Iterator<Item = Result<Token, SyntaxError>> + '_ {
    let mut pos = TextSize::from(0);

    iter::from_fn(move || {
        if input.is_empty() {
            return None;
        }

        match first_token(input) {
            Ok(token) => {
                let len = token.len;

                let token = Token::new(token.inner, TextRange::at(pos, len));

                pos += len;
                input = &input[len.into()..];

                Some(Ok(token))
            }
            Err(err) => {
                let len = err.len;

                let e = SyntaxError::new(err.inner.0, TextRange::at(pos, len));

                pos += len;
                input = &input[len.into()..];

                Some(Err(e))
            }
        }
    })
}

fn first_token(input: &str) -> Result<WithLen<SyntaxKind>, WithLen<LexErrorMsg>> {
    Lexer::new(input).next_token()
}

#[derive(Debug, Error)]
#[error("{msg}")]
pub struct LexError {
    pub msg: LexErrorMsg,
    pub range: TextRange,
}

impl LexError {
    pub fn to_unknown(&self) -> Token {
        Token {
            kind: T![unknown],
            range: self.range,
        }
    }
}

#[derive(Debug, Error, PartialEq, Eq)]
#[error("{0}")]
pub struct LexErrorMsg(String);

type LexResult<T, E = LexErrorMsg> = Result<T, E>;

pub struct Lexer<'a> {
    input_len: u32,
    chars: Chars<'a>,
}

impl<'a> Lexer<'a> {
    fn new(input: &'a str) -> Lexer<'a> {
        Lexer {
            input_len: input.len() as u32,
            chars: input.chars(),
        }
    }

    fn chars(&self) -> Chars<'a> {
        self.chars.clone()
    }

    fn nth(&self, n: u32) -> char {
        self.chars().nth(n as usize).unwrap_or(EOF_CHAR)
    }

    fn eof(&self) -> bool {
        self.at('\0')
    }

    /// Peeks next char from stream without consuming it
    fn current(&self) -> char {
        self.nth(0)
    }

    fn at<T: Peek>(&self, t: T) -> bool {
        T::peek(t, self)
    }

    /// Checks if there is nothing more to consume.
    fn is_eof(&self) -> bool {
        self.chars.as_str().is_empty()
    }

    fn chars_len(&self) -> u32 {
        self.chars.as_str().len() as u32
    }

    /// Returns amount of already consumed symbols.
    fn pos(&self) -> u32 {
        self.input_len - self.chars_len()
    }

    fn bump(&mut self) -> Option<char> {
        self.chars.next()
    }

    fn bump_then(&mut self) -> char {
        self.bump();
        self.current()
    }

    fn next_token(mut self) -> Result<WithLen<SyntaxKind>, WithLen<LexErrorMsg>> {
        self.lex_main()
            .map(|kind| WithLen::new(kind, self.pos().into()))
            .map_err(|e| WithLen::new(e, self.pos().into()))
    }

    fn accept<T: Accept + Copy>(&mut self, t: T) -> bool {
        T::accept(t, self)
    }

    fn accept_while<T: Accept + Copy>(&mut self, t: T) {
        T::accept_while(t, self)
    }

    fn accept_while_count<T: Accept + Copy>(&mut self, t: T) -> u32 {
        T::accept_while_count(t, self)
    }

    fn accept_repeat<T: Accept + Copy>(&mut self, t: T, repeat: u32) -> bool {
        T::accept_repeat(t, self, repeat)
    }

    fn lex_main(&mut self) -> LexResult<SyntaxKind> {
        let c = self.current();

        // return on special cases
        let kind = match c {
            '=' => match self.bump_then() {
                '=' => {
                    self.bump();
                    done!(T![==])
                }
                _ => done!(T![=]),
            },
            '~' => match self.bump_then() {
                '=' => {
                    self.bump();
                    done!(T![~=]);
                }
                _ => done!(T![unknown]),
            },

            '(' => T!['('],
            ')' => T![')'],
            '{' => T!['{'],
            '}' => T!['}'],
            '[' => match self.bump_then() {
                '[' => done!(self.multiline_string()?),
                '=' => done!(self.multiline_string()?),
                _ => done!(T!['[']),
            },
            ']' => T![']'],

            ',' => T![,],
            '.' => match self.bump_then() {
                '.' => match self.bump_then() {
                    '.' => {
                        self.bump();
                        done!(T![...])
                    }
                    _ => done!(T![..]),
                },
                _ => done!(T![.]),
            },
            ';' => T![;],
            ':' => match self.bump_then() {
                ':' => {
                    self.bump();
                    done!(T![::]);
                }
                _ => done!(T![:]),
            },

            '+' => T![+],
            '*' => T![*],
            '/' => T![/],
            '%' => T![%],
            '^' => T![^],

            '!' => T![!],

            '-' => match self.bump_then() {
                '-' => done!(self.comment()),
                _ => done!(T![-]),
            },

            c @ '\'' | c @ '"' => done!(self.string(c)?),

            _ if is_ident_start(c) => done!(self.ident()),
            _ if is_number(c) => done!(self.number()?),
            _ if is_whitespace(c) => done!(self.whitespace()),

            _ => T![unknown],
        };

        // if we got here, that means that the token was only length 1
        self.bump().unwrap();

        Ok(kind)
    }

    fn number(&mut self) -> LexResult<SyntaxKind> {
        assert!(self.at(is_number));

        if self.accept(('0', 'x')) {
            self.accept_while(is_hex);
            done!(T![number]);
        }

        self.accept_while(is_number);
        if self.accept('.') {
            self.accept_while(is_number);
        }
        if self.accept('e') || self.accept('E') {
            self.accept('-');
            self.accept_while(is_number);
        }

        Ok(T![number])
    }

    fn whitespace(&mut self) -> SyntaxKind {
        assert!(self.at(is_whitespace));

        self.accept_while(is_whitespace);
        T![whitespace]
    }

    fn comment(&mut self) -> SyntaxKind {
        assert!(self.at('-'));

        self.bump().unwrap();

        self.chars.find(|c| *c == '\n');

        T![comment]
    }

    fn bracket_enclosed(&mut self) -> LexResult<()> {
        fn close(l: &mut Lexer<'_>, count: u32) -> LexResult<()> {
            let mut err = Ok(());

            let mut set_err = || {
                if err.is_ok() {
                    err = Err(LexErrorMsg("Invalid bracket notation".to_string()));
                }
            };

            macro_rules! expect {
                ($it:expr) => {
                    if !$it {
                        set_err()
                    }
                };
            }

            expect!(l.accept('['));

            l.accept_while(|c| c != ']');

            expect!(l.accept(']'));

            let close_count = l.accept_while_count('=');
            if count != close_count {
                set_err();
                l.accept_while('=');
            }

            expect!(l.accept(']'));

            err
        }

        assert_matches!(self.current(), '[' | '=');

        if self.at('[') {
            close(self, 0)
        } else if self.at('=') {
            let count = self.accept_while_count('=');
            close(self, count)
        } else {
            Ok(())
        }
    }

    fn multiline_string(&mut self) -> LexResult<SyntaxKind> {
        assert!(self.at('[') || self.at('='));

        self.bracket_enclosed().map(|()| T![str])
    }

    fn multiline_comment(&mut self) -> LexResult<SyntaxKind> {
        assert!(self.at('[') || self.at('='));

        self.bracket_enclosed().map(|()| T![comment])
    }

    fn string(&mut self, delimit: char) -> LexResult<SyntaxKind> {
        assert!(matches!(delimit, '\'' | '"'));
        assert_eq!(self.current(), delimit);
        self.bump().unwrap();

        self.chars
            .find(|c| *c == delimit)
            .ok_or(LexErrorMsg("UnfinishedString".into()))?;

        Ok(T![str])
    }

    fn ident(&mut self) -> SyntaxKind {
        assert!(self.at(is_ident_start));

        let start = self.pos();
        let text = self.chars.as_str();

        self.bump().unwrap();
        self.accept_while(is_ident_continue);

        let text = &text[0..(self.pos() - start) as usize];
        SyntaxKind::from_keyword(text).unwrap_or(T![ident])
    }
}

pub trait Peek {
    fn nth(self, p: &Lexer<'_>, n: u32) -> bool;

    fn peek(self, p: &Lexer<'_>) -> bool
    where
        Self: Sized,
    {
        self.nth(p, 0)
    }
}

trait Accept: Peek {
    fn accept(self, l: &mut Lexer<'_>) -> bool
    where
        Self: Peek + Sized,
    {
        if self.peek(l) && !l.eof() {
            l.bump().unwrap();
            true
        } else {
            false
        }
    }

    fn accept_while(self, l: &mut Lexer<'_>)
    where
        Self: Peek + Sized + Copy,
    {
        while self.accept(l) {}
    }

    fn accept_while_count(self, l: &mut Lexer<'_>) -> u32
    where
        Self: Peek + Sized + Copy,
    {
        let mut count = 0;
        while self.accept(l) {
            count += 1;
        }
        count
    }

    fn accept_repeat(self, l: &mut Lexer<'_>, repeat: u32) -> bool
    where
        Self: Peek + Sized + Copy,
    {
        if (0..repeat).all(|i| self.nth(l, i)) {
            for _ in 0..repeat {
                l.bump().unwrap();
            }
            true
        } else {
            false
        }
    }
}

impl Peek for char {
    fn nth(self, l: &Lexer<'_>, n: u32) -> bool {
        l.nth(n) == self
    }
}
impl Accept for char {}

impl<T: Fn(char) -> bool> Peek for T {
    fn nth(self, l: &Lexer<'_>, n: u32) -> bool {
        self(l.nth(n))
    }
}
impl<T: Fn(char) -> bool> Accept for T {}

impl Peek for Range<char> {
    fn nth(self, l: &Lexer<'_>, n: u32) -> bool {
        self.contains(&l.nth(n))
    }
}
impl Accept for Range<char> {}

macro_rules! tuple_impls {
    { $( ($( $n:tt $name:ident )+ $(,)?) ),+ $(,)? } => {
        $(
            impl<$($name),+> Peek for ($($name),+)
            where
                $($name: Peek,)+
                {
                    fn nth(self, l: &Lexer<'_>, n: u32) -> bool {
                        $( self.$n.nth(l, $n + n) )&&+
                    }
                }

            impl<$($name),+> Accept for ($($name),+)
            where
                $($name: Accept,)+
                {
                    fn accept(self, l: &mut Lexer<'_>) -> bool {
                        if self.peek(l) {
                            $( $n; l.bump().unwrap(); )+
                            true
                        } else {
                            false
                        }
                    }
                }
        )+
    };
}

tuple_impls! {
    (0 TO 1 T1),
    (0 TO 1 T1 2 T2),
    (0 TO 1 T1 2 T2 3 T3),
    (0 TO 1 T1 2 T2 3 T3 4 T4),
    (0 TO 1 T1 2 T2 3 T3 4 T4 5 T5),
    (0 TO 1 T1 2 T2 3 T3 4 T4 5 T5 6 T6),
    (0 TO 1 T1 2 T2 3 T3 4 T4 5 T5 6 T6 7 T7),
    (0 TO 1 T1 2 T2 3 T3 4 T4 5 T5 6 T6 7 T7 8 T8),
}

const fn is_number(c: char) -> bool {
    matches!(c, '0'..='9')
}

const fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

const fn is_ident_continue(c: char) -> bool {
    is_ident_start(c) || matches!(c, '0'..='9')
}

const fn is_whitespace(c: char) -> bool {
    c.is_ascii_whitespace()
}

const fn is_hex(c: char) -> bool {
    matches!(c, '0'..='9' | 'A'..='F' | 'a'..='f')
}

#[cfg(test)]
mod tests {
    use std::str;

    use super::*;

    #[test]
    fn accept_tuple() {
        let mut lexer = Lexer::new("[=");
        lexer.accept(('[', '='));
        assert!(lexer.eof());
    }

    #[test]
    fn accept_tuple_fail() {
        let mut lexer = Lexer::new("[]");
        lexer.accept(('[', '='));
        assert_eq!(lexer.pos(), 0);
        assert_eq!(lexer.current(), '[');
        lexer.bump().unwrap();
        assert_eq!(lexer.current(), ']');
    }

    #[test]
    fn accept_repeat() {
        let mut lexer = Lexer::new("===================");
        lexer.accept_repeat('=', 19);
        assert!(lexer.eof());
    }

    #[test]
    fn accept_repeat_not_enough() {
        let mut lexer = Lexer::new("==");
        lexer.accept_repeat('=', 4);
        assert_eq!(lexer.pos(), 0);
        lexer.accept_repeat('=', 2);
        assert!(lexer.eof());
    }

    #[test]
    fn accept_repeat_none() {
        let mut lexer = Lexer::new("not");
        assert!(lexer.accept_repeat('=', 0));
        assert_eq!(lexer.pos(), 0);
    }
}
