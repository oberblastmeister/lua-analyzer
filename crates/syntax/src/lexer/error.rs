use crate::{SyntaxError, SyntaxKind};

/// A Result that is always ok and can have an optional error
pub struct OkResult<T, E>(T, Option<E>);

impl<T, E> OkResult<T, E> {
    pub fn new(t: T, err: Option<E>) -> OkResult<T, E> {
        OkResult(t, err)
    }

    pub fn just(t: T) -> OkResult<T, E> {
        OkResult(t, None)
    }

    pub fn map<U, F>(self, f: F) -> OkResult<U, E>
    where
        F: FnOnce(T) -> U,
    {
        OkResult(f(self.0), self.1)
    }

    pub fn map_err<F, U>(self, f: F) -> OkResult<T, U>
    where
        F: FnOnce(E) -> U,
    {
        OkResult(self.0, self.1.map(f))
    }

    pub fn inner(self) -> (T, Option<E>) {
        (self.0, self.1)
    }

    pub fn ok(self) -> T {
        self.0
    }

    pub fn ok_ref(&self) -> &T {
        &self.0
    }

    pub fn err(self) -> Option<E> {
        self.1
    }

    pub fn has_err(&self) -> bool {
        self.1.is_some()
    }
}

impl<E> OkResult<(), E> {
    pub fn unit(err: Option<E>) -> OkResult<(), E> {
        OkResult((), err)
    }

    pub fn empty() -> OkResult<(), E> {
        OkResult((), None)
    }
}

// impl<T, E> ResultInfo<T, E> {
//     pub fn just(kind: T) -> ResultInfo<SyntaxKind, E> {
//         ResultInfo(kind, None)
//     }
// }

pub type LexResult<T> = OkResult<T, String>;

pub type SyntaxResult<T> = OkResult<T, SyntaxError>;
