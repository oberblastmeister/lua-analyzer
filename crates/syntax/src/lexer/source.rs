use std::convert::TryFrom;
use std::str::Chars;

use accept::Advancer;

pub(crate) const EOF_CHAR: char = '\0';

pub struct CharSource<'a> {
    input_len: u32,
    chars: Chars<'a>,
}

impl<'a> CharSource<'a> {
    pub fn new(text: &'a str) -> CharSource<'a> {
        CharSource { input_len: u32::try_from(text.len()).unwrap(), chars: text.chars() }
    }

    pub fn chars(&self) -> Chars<'a> {
        self.chars.clone()
    }

    pub fn chars_len(&self) -> u32 {
        self.chars.as_str().len() as u32
    }

    /// Returns amount of already consumed chars.
    pub fn pos(&self) -> u32 {
        self.input_len - self.chars_len()
    }

    /// The rest of the str that hasn't been consumed yet
    pub fn rest(&self) -> &'a str {
        self.chars.as_str()
    }
}

impl Advancer for CharSource<'_> {
    type Item = char;

    fn advance(&mut self) -> Option<char> {
        self.chars.next()
    }

    fn nth(&self, n: u32) -> char {
        self.chars().nth(n as usize).unwrap_or(EOF_CHAR)
    }

    fn is_eof(&self) -> bool {
        self.nth_is_eof(0)
    }

    fn nth_is_eof(&self, n: u32) -> bool {
        self.chars.clone().nth(n as usize).is_none()
    }
}
