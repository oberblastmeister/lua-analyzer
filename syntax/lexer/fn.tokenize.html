<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../syntax/fn.tokenize.html">
</head>
<body>
    <p>Redirecting to <a href="../../syntax/fn.tokenize.html">../../syntax/fn.tokenize.html</a>...</p>
    <script>location.replace("../../syntax/fn.tokenize.html" + location.search + location.hash);</script>
</body>
</html>