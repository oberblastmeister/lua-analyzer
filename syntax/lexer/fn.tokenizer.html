<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../syntax/fn.tokenizer.html">
</head>
<body>
    <p>Redirecting to <a href="../../syntax/fn.tokenizer.html">../../syntax/fn.tokenizer.html</a>...</p>
    <script>location.replace("../../syntax/fn.tokenizer.html" + location.search + location.hash);</script>
</body>
</html>