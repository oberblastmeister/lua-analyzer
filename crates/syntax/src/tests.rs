use super::lexer::{tokenize, tokenize_iter};
use super::*;
use std::{
    fmt, fs,
    path::{Path, PathBuf},
    str,
};

#[test]
fn parser() {
    dir_tests(snapshots_dir(), &["parser/ok"], |path, text| {
        let parse = SourceFile::parse(text);
        let errors = parse.errors();
        assert_errors_are_absent(errors, path);
        dump_parse(parse)
    });

    dir_tests(snapshots_dir(), &["parser/err"], |path, text| {
        let parse = SourceFile::parse(text);
        let errors = parse.errors();
        assert_errors_are_present(errors, path);
        dump_parse(parse)
    })
}

#[test]
fn lexer() {
    dir_tests(snapshots_dir(), &["lexer/ok"], |path, text| {
        let (tokens, errors) = tokenize(text);
        assert_errors_are_absent(&errors, path);
        assert_unknowns_are_absent(&tokens, path);
        format!("{:#?}", tokens)
    });

    dir_tests(snapshots_dir(), &["lexer/err"], |path, text| {
        let (tokens, errors) = tokenize(text);
        assert_errors_or_unknowns_are_present(&errors, &tokens, path);
        format!("{:#?}\n\n{:#?}", tokens, errors)
    })
}

fn dir_tests<P, F>(test_data_dir: P, paths: &[&str], f: F)
where
    P: AsRef<Path>,
    F: Fn(&Path, &str) -> String,
{
    for (path, input_code) in collect_lua_files(test_data_dir.as_ref(), paths) {
        assert_path(&path, &f(&path, &input_code))
    }
}

fn assert_path<P>(path: P, actual: &str)
where
    P: AsRef<Path>,
{
    let path = path.as_ref();

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_path(
        path.parent()
            .expect("There is no parent for the given path"),
    );
    settings.set_prepend_module_to_snapshot(false);
    settings.set_input_file(path);
    settings.bind(|| {
        let name = path
            .file_stem()
            .expect("No file stem")
            .to_str()
            .expect("Could not turn path to str");
        insta::assert_snapshot!(name, actual);
    })
}

fn collect_lua_files(root_dir: &Path, paths: &[&str]) -> Vec<(PathBuf, String)> {
    paths
        .iter()
        .flat_map(|path| {
            let path = root_dir.join(path);
            lua_files_in_dir(&path).into_iter()
        })
        .map(|path| {
            let text = read_text(&path);
            (path, text)
        })
        .collect()
}

fn lua_files_in_dir(dir: &Path) -> Vec<PathBuf> {
    let mut acc = Vec::new();
    for file in fs::read_dir(&dir).unwrap() {
        let file = file.unwrap();
        let path = file.path();
        if path.extension().unwrap_or_default() == "lua" {
            acc.push(path);
        }
    }
    acc.sort();
    acc
}
fn assert_errors_or_unknowns_are_present(errors: &[SyntaxError], tokens: &[Token], path: &Path) {
    assert!(
        !errors.is_empty() || !tokens.is_empty(),
        "There should be errors in the file {:?}",
        path.display()
    );
}

fn assert_errors_are_present(errors: &[SyntaxError], path: &Path) {
    assert!(
        !errors.is_empty(),
        "There should be errors in the file {:?}",
        path.display()
    );
}

fn assert_errors_are_absent(errors: &[SyntaxError], path: &Path) {
    assert_eq!(
        errors,
        &[] as &[SyntaxError],
        "There should be no errors in the file {:?}",
        path.display(),
    );
}

fn collect_unknowns(tokens: &[Token]) -> Vec<Token> {
    tokens
        .into_iter()
        .filter(|t| t.kind == T![unknown])
        .map(|t| *t)
        .collect()
}

fn assert_unknowns_are_absent(tokens: &[Token], path: &Path) {
    let unknowns = collect_unknowns(tokens);
    assert_eq!(
        unknowns,
        &[] as &[Token],
        "There should be no unknowns in the file {:?}",
        path.display()
    )
}

fn assert_unknowns_are_present(tokens: &[Token], path: &Path) {
    let unknowns = collect_unknowns(tokens);
    assert!(
        !unknowns.is_empty(),
        "There should be unknowns in the file {:?}",
        path.display()
    )
}

fn dump_parse<T: AstNode + fmt::Debug>(parse: Parse<T>) -> String {
    let mut s = String::new();
    s.push_str(&format!("{:#?}", parse.syntax_node()));
    s.push_str(&format_errors(parse.errors()));
    s
}

fn dump_parse_no_errors(parse: Parse<SourceFile>) -> Result<String, String> {
    if !parse.errors().is_empty() {
        return Err(dump_parse(parse));
        // panic!(
        //     "Should not have any errors {:#?}\n{}",
        //     parse.syntax_node(),
        //     format_errors(parse.errors())
        // )
    }
    Ok(parse.debug_dump())
}

/// Read file and normalize newlines.
///
/// `rustc` seems to always normalize `\r\n` newlines to `\n`:
///
/// ```
/// let s = "
/// ";
/// assert_eq!(s.as_bytes(), &[10]);
/// ```
///
/// so this should always be correct.
fn read_text(path: &Path) -> String {
    fs::read_to_string(path)
        .unwrap_or_else(|_| panic!("File at {:?} should be valid", path))
        .replace("\r\n", "\n")
}

fn snapshots_dir() -> PathBuf {
    project_root().join("crates/syntax/snapshots")
}

fn project_root() -> PathBuf {
    let dir = env!("CARGO_MANIFEST_DIR");
    PathBuf::from(dir)
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .to_owned()
}

#[test]
fn nothing() {
    insta::assert_debug_snapshot!(SourceFile::parse("").syntax_node(), @"SourceFile@0..0")
}

#[test]
fn fuzz1() {
    SourceFile::parse("e,,");
}

#[test]
fn local_only() {
    SourceFile::parse("local");
}
