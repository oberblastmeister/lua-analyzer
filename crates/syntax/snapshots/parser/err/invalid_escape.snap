---
source: crates/syntax/src/tests.rs
expression: actual
input_file: crates/syntax/snapshots/parser/err/invalid_escape.lua

---
SourceFile@0..37
  LocalAssignStmt@0..36
    LocalKw@0..5 "local"
    Whitespace@5..6 " "
    MultiName@6..7
      Name@6..7
        Ident@6..7 "s"
    Whitespace@7..8 " "
    Eq@8..9 "="
    Whitespace@9..10 " "
    MultivalExpr@10..36
      Literal@10..36
        Str@10..36 "\"asdf\\w\\u\\oasd\\q\\z\\a  ..."
  Whitespace@36..37 "\n"

=============================
Errors:
=============================
[
    SyntaxError(
        "Unexpected newline in string",
        10..36,
    ),
    SyntaxError(
        "Invalid escape sequence",
        15..17,
    ),
    SyntaxError(
        "Invalid escape sequence",
        17..19,
    ),
    SyntaxError(
        "Invalid escape sequence",
        19..21,
    ),
    SyntaxError(
        "Invalid escape sequence",
        24..26,
    ),
    SyntaxError(
        "Invalid escape sequence",
        26..28,
    ),
    SyntaxError(
        "Invalid escape sequence",
        34..35,
    ),
]
